lucene/grouping/src/test/org/apache/lucene/search/grouping/DocValuesPoolingReaderTest.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+public class DocValuesPoolingReaderTest extends LuceneTestCase {
+  
+  private static RandomIndexWriter w;
+  private static Directory dir;
+  private static DirectoryReader reader;
+
+  @BeforeClass
+  public static void index() throws IOException {
+    dir = newDirectory();
+    w = new RandomIndexWriter(
+        random(),
+        dir,
+        newIndexWriterConfig(new MockAnalyzer(random())).setMergePolicy(newLogMergePolicy()));
+    Document doc = new Document();
+    doc.add(new BinaryDocValuesField("bin", new BytesRef("binary")));
+    doc.add(new BinaryDocValuesField("bin2", new BytesRef("binary2")));
+    
+    doc.add(new NumericDocValuesField("num", 1L));
+    doc.add(new NumericDocValuesField("num2", 2L));
+    
+    doc.add(new SortedNumericDocValuesField("sortnum", 3L));
+    doc.add(new SortedNumericDocValuesField("sortnum2", 4L));
+    
+    doc.add(new SortedDocValuesField("sort",  new BytesRef("sorted")));
+    doc.add(new SortedDocValuesField("sort2",  new BytesRef("sorted2")));
+    
+    doc.add(new SortedSetDocValuesField("sortset", new BytesRef("sortedset")));
+    doc.add(new SortedSetDocValuesField("sortset2", new BytesRef("sortedset2")));
+    
+    w.addDocument(doc);
+    w.commit();
+    reader = w.getReader();
+    w.close();
+  }
+  
+  public void testDVCache() throws IOException {
+    assertFalse(reader.leaves().isEmpty());
+    for (LeafReaderContext leaf : reader.leaves()) {
+      final DocValuesPoolingReader caching = new DocValuesPoolingReader(leaf.reader());
+      
+      assertSame(assertBinaryDV(caching, "bin", "binary"), 
+          caching.getBinaryDocValues("bin"));
+      assertSame(assertBinaryDV(caching, "bin2", "binary2"), 
+          caching.getBinaryDocValues("bin2"));
+      
+      assertSame(assertNumericDV(caching, "num", 1L), 
+          caching.getNumericDocValues("num"));
+      assertSame(assertNumericDV(caching, "num2", 2L), 
+          caching.getNumericDocValues("num2"));
+      
+      assertSame(assertSortedNumericDV(caching, "sortnum", 3L), 
+          caching.getSortedNumericDocValues("sortnum"));
+      assertSame(assertSortedNumericDV(caching, "sortnum2", 4L), 
+          caching.getSortedNumericDocValues("sortnum2"));
+      
+      assertSame(assertSortedDV(caching, "sort", "sorted"), 
+          caching.getSortedDocValues("sort"));
+      assertSame(assertSortedDV(caching, "sort2", "sorted2"), 
+          caching.getSortedDocValues("sort2"));
```

Comment:
I think this still doesn't test iteration through a single doc's values on two instances of the shared DV?  We need to pull the iterator twice, advance both to the same doc, and then iterate through the values on both - as I read it, currently if you iterate through the values on one, then you can't iterate through them on the other because the within-document iterator is shared between the two of them.  We need to cache the values.

Reply:
Hi, @romseygeek , thanks for feedback. I agree to add more tests. I suppose such functionality is not necessary. Docs arrive into grouping collector in-order. Thus, different group heads might read vals one by one also in-order. When it needs to jump back and read cached values? 
Currently, test coverage is shallow, but it proves this read-once idea.